### ISSUE TITLE
[MNG-7010] Omit "NB: JAVA_HOME should point to a JDK not a JRE" except when that is the problem

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8026]
**[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo)** opened **[MNG-7010](https://issues.apache.org/jira/browse/MNG-7010?redirect=false)** and commented

Got this message a bunch today:

 

$ mvn verify
The JAVA_HOME environment variable is not defined correctly
This environment variable is needed to run this program
NB: JAVA_HOME should point to a JDK not a JRE

 

fair enough. My JAVA_HOME is broken, for reasons. However the problem is **not** that it points to a JDK instead of a JRE. Maven should omit "NB: JAVA_HOME should point to a JDK not a JRE" except when that is the problem.

Also, I have no idea what NB means here. Nota bene perhaps?

 

 

---

**Remote Links:**
- [GitHub Pull Request #512
](https://github.com/apache/maven/pull/512)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.2](https://github.com/apache/maven/milestone/103?closed=1)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Frankly,  I took this from Tomcat startup script. Maybe [Mark Thomas](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=markt) can enlighten us?

**Comment by jira-importer** [comment]
**[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo)** commented

Yes, that makes sense. Speciifcally:

1. Warn "The JAVA_HOME environment variable is not defined correctly
   This environment variable is needed to run this program." when the JAVA_HOME environment variable is not defined correctly.

2. Drop "NB" completely. It's very obscure.

3. If and only if the problem is that JAVA_HOME should points to a JRE instead of a JDK, then also add, "JAVA_HOME should point to a JDK, not a JRE."

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Makes sense..

Do you agree with the testing for `javac`?

**Comment by jira-importer** [comment]
**[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo)** commented

Looks like that would work. I don't know if there's a more official way to do this:

$ ls /Library/Java/JavaVirtualMachines/jdk1.8.0_172.jdk/Contents/Home/bin
appletviewer	javah		jjs		jvisualvm	schemagen
extcheck	javap		jmap		keytool		serialver
idlj		javapackager	jmc		native2ascii	servertool
jar		jcmd		jps		orbd		tnameserv
jarsigner	jconsole	jrunscript	pack200		unpack200
java		jdb		jsadebugd	policytool	wsgen
javac		jdeps		jstack		rmic		wsimport
javadoc		jhat		jstat		rmid		xjc
javafxpackager	jinfo		jstatd		rmiregistry
(base) ~$ ls /Library/Java/JavaVirtualMachines/jdk1.8.0_172.jdk/Contents/Home/jre
COPYRIGHT				THIRDPARTYLICENSEREADME.txt
LICENSE					Welcome.html
README					bin
THIRDPARTYLICENSEREADME-JAVAFX.txt	lib
$ ls /Library/Java/JavaVirtualMachines/jdk1.8.0_172.jdk/Contents/Home/jre/bin
java		orbd		rmid		tnameserv
jjs		pack200		rmiregistry	unpack200
keytool		policytool	servertool

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Actually, I think we should drop the line completely.
To run Maven you will need `java`.
If one wants to compile java code, you will need `javac`, which is a requirement for the maven-compiler-plugin and not Maven.
If one wants to generate java documentation, you will need `javadoc`, which is a requirement for the maven-javadoc-plugin and not Maven.
When you are using toolchains, or have the path to the JDK specified otherwise, JRE is enough to run Maven.
It it is not a Java project, JRE is enough to run Maven.

I think the plugins already do these checks, otherwise they should, but having a JDK instead or JRE doesn't have to be a Maven requirement.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I do agree with [Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), therefore I would skim the PR and change https://github.com/apache/maven/blob/a2ab329edc51c1f59493babc4f0f3d4b246a0c36/apache-maven/src/assembly/shared/init#L19-L37 to `java` instead of `javac`.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

We should improve the message: if JAVA_HOME is set, but %JAVA_HOME%/bin/java doesn't exist, we should say so (just hit another SO question where JAVA_HOME included the bin directory...)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

So you'd change

```
147   echo "The JAVA_HOME environment variable is not defined correctly" >&2
148   echo "This environment variable is needed to run this program" >&2
```

?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

It is not needed anymore to set JAVA_HOME.
If it is set, %JAVA_HOME%/bin/java should be used to start Maven, otherwise java should be picked up from the PATH.
This should be explained on the [install page](https://maven.apache.org/install.html).

With this we can assume that JAVA_HOME was set on purpose, but misconfigured.
I suggest to have a message like:

```
The JAVA_HOME environment variable is not defined correctly, so Apache Maven cannot be started.
JAVA_HOME is set to "<actual value>", but "%JAVA_HOME%/bin/java" does not exist.
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Created MNGSITE-456. [Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), I will create a new issue with your script improvement and finish this one with the remove of the mentioned line in the summary in 4.0.x and 3.8.x.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [5fa834f78be178b4f40b5ffda6d3c0dc873b497e](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=5fa834f78be178b4f40b5ffda6d3c0dc873b497e) and [f32eb09892686cba083f9f175b3c8eac0ac42ae7](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=f32eb09892686cba083f9f175b3c8eac0ac42ae7) for `maven-3.8.x` branch.
